### PR TITLE
Allow guessing of Proxy Authentication

### LINKED ISSUE #125: Guess Proxy Authentication
I am developing a 'GuessAuthProxyAuth' class for a project I'm working on that subclasses GuessAuth, and allows for guessing of both  HTTP authentication as well as Proxy Authentication.  Essentially, it does for Proxy authentication methods what GuessAuth does for HTTP authentication.

I would like to be able to contribute it here, since I feel others could potentially benefit from it.

### PR DESCRIPTION
**Pull request by pcreech**
Allows the guessing of proxy authentication alongside guessing
of HTTP Authentication

Closes #125

### CODE REVIEW

**Comment by Lukasa** [comment]
Seems reasonable to me! Thanks @pcreech!

File: docs/authentication.rst [hunk 1]
```diff
@@ -87,6 +87,36 @@ require authentication, no credentials are sent.
 .. autoclass:: requests_toolbelt.auth.guess.GuessAuth
 
 
+GuessAuthProxyAuth
+---------
```
Comment (sigmavirus24) [inline docs/authentication.rst]: Please match the number of `-`s to the length of the title.

Comment (sigmavirus24) [inline docs/authentication.rst]: Let's also make the name simpler, e.g., `GuessProxyAuth`.

File: docs/authentication.rst [hunk 2]
```diff
@@ -87,6 +87,36 @@ require authentication, no credentials are sent.
...
+
+The :class:`~requests_toolbelt.auth.guess.GuessAuthProxyAuth` authentication class
+provides GuessAuth, while also automatically detecting whether to use basic auth
```
Comment (sigmavirus24) [inline docs/authentication.rst]: Let's reword this to something like

```rst
The :class:`~requests_toolbelt.auth.guess.GuessProxyAuth` handler will automatically detect whether to use basic authentication or digest authentication when authenticating to the provided proxy.
```

File: docs/authentication.rst [hunk 3]
```diff
@@ -87,6 +87,36 @@ require authentication, no credentials are sent.
...
+or digest auth with a provided proxy.
+
+.. code-block:: python
+
+    import requests
+    from requests_toolbelt.auth.guess import GuessAuthProxyAuth
+
+    proxies = {
+        "http": "http://PROXYSERVER:PROXYPORT",
+        "https": "https://PROXYSERVER:PROXYPORT",
```
Comment (sigmavirus24) [inline docs/authentication.rst]: The correct way to do this in requests is typically:

```
"https": "http://PROXYSERVER:PROXYPORT",
```

Because of issues surrounding how we talk to HTTPS sites over proxies run on HTTPS. Since we recommend the above in requests, our examples here should do the same.

File: docs/authentication.rst [hunk 4]
```diff
@@ -87,6 +87,36 @@ require authentication, no credentials are sent.
...
+    }
+    requests.get('http://httpbin.org/basic-auth/user/passwd',
+                 auth=GuessAuth('user', 'passwd', 'proxyusr', 'proxypass'),
```
Comment (sigmavirus24) [inline docs/authentication.rst]: Should this be `GuessProxyAuth`?

File: docs/authentication.rst [hunk 5]
```diff
@@ -87,6 +87,36 @@ require authentication, no credentials are sent.
...
+---------
+
+The :class:`~requests_toolbelt.auth.guess.GuessAuthProxyAuth` authentication class
+provides GuessAuth, while also automatically detecting whether to use basic auth
+or digest auth with a provided proxy.
+
+.. code-block:: python
+
+    import requests
+    from requests_toolbelt.auth.guess import GuessAuthProxyAuth
```
Comment (sigmavirus24) [inline docs/authentication.rst]: Don't forget to change this to `GuessProxyAuth`

File: docs/authentication.rst [hunk 6]
```diff
@@ -87,6 +87,36 @@ require authentication, no credentials are sent.
...
+
+    proxies = {
+        "http": "http://PROXYSERVER:PROXYPORT",
+        "https": "https://PROXYSERVER:PROXYPORT",
+    }
+    requests.get('http://httpbin.org/basic-auth/user/passwd',
+                 auth=GuessAuth('user', 'passwd', 'proxyusr', 'proxypass'),
+                 proxies=proxies)
+    requests.get('http://httpbin.org/digest-auth/auth/user/passwd',
+                 auth=GuessAuth('user', 'passwd', 'proxyusr', 'proxypass'),
```
Comment (sigmavirus24) [inline docs/authentication.rst]: Same here

**Comment by sigmavirus24** [comment]
We can't use Betamax for these tests because they don't record the proxy interactions. We'll need to use mock objects to test this code.

**Comment by pcreech** [comment]
Cleaned up the names and docs.

I left the GuessAuth testing in there as is, to ensure it preserves the functionality of GuessAuth without a proxy through GuessProxyAuth.

I'm looking into a way to mock/test the proxy piece of it.  I'm also open to any suggestings

File: requests_toolbelt/auth/guess.py [hunk 7]
```diff
@@ -77,3 +77,72 @@ def __call__(self, request):
...
+
+
+class GuessProxyAuth(GuessAuth):
+    """
+    Guesses the auth type by WWW-Authentication and Proxy-Authentication
+    headers
+    """
+    def __init__(self, username=None, password=None,
+                 proxy_username=None, proxy_password=None):
+        GuessAuth.__init__(self, username, password)
```
Comment (sigmavirus24) [inline requests_toolbelt/auth/guess.py]: Why not use super here? The library doesn't support python 2.5 and as far as I remember, `AuthBase` from requests inherits from `object` so there should be no worry about being able to use super in this case.

**Comment by sigmavirus24** [comment]
I personally prefer simply using the mock library. I'm not sure about a strategy for using this because I'd have to think a bit on where the right place to mock this out would be. That said, if you're uncomfortable with mock, I can write the tests for this. Just get rid of the betamax tests and their cassettes and I'll take care of the rest.

**Comment by pcreech** [comment]
Sounds good, i don't mind doing the mocking.  Need the experience.  Should have some updates again soon!

**Comment by sigmavirus24** [comment]
@pcreech There's absolutely no rush from my end. Please ping me on this *when* you've pushed updates. (This is just a reminder that GitHub doesn't notify me when you've pushed new commits that should be reviewed although I really wish they would do that somehow.)

**Comment by pcreech** [comment]
@sigmavirus24 I've updated the unit tests.

File: requests_toolbelt/auth/guess.py [hunk 8]
```diff
@@ -77,3 +77,72 @@ def __call__(self, request):
...
+
+        try:
+            self.auth.init_per_thread_state()
+        except AttributeError:
+            pass
+
+        return self.proxy_auth.handle_407(r, **kwargs)
+
+    def handle_407(self, r, **kwargs):
+
```
Comment (sigmavirus24) [inline requests_toolbelt/auth/guess.py]: Please remove the new-line here.

File: requests_toolbelt/auth/guess.py [hunk 9]
```diff
@@ -77,3 +77,72 @@ def __call__(self, request):
...
+        proxy_authenticate = r.headers.get('Proxy-Authenticate', '').lower()
+
+        if 'basic' in proxy_authenticate:
+            return self._handle_basic_auth_407(r, kwargs)
+
+        if 'digest' in proxy_authenticate:
+            return self._handle_digest_auth_407(r, kwargs)
+
+    def __call__(self, request):
+
```
Comment (sigmavirus24) [inline requests_toolbelt/auth/guess.py]: Please remove the new-line here.

File: tests/test_auth.py [hunk 10]
```diff
@@ -44,3 +45,36 @@ def test_no_auth(self):
...
+
+
+class TestGuessProxyAuth(unittest.TestCase):
+
+    @mock.patch('requests_toolbelt.auth.http_proxy_digest.HTTPProxyDigestAuth.handle_407')
+    def test_handle_407_header_digest(self, mock_handle_407):
+        r = requests.Response()
+        r.headers['Proxy-Authenticate'] = 'Digest nonce="d2b19757d3d656a283c99762cbd1097b", opaque="1c311ad1cc6e6183b83bc75f95a57893", realm="me@kennethreitz.com", qop=auth'
+
+        guess_auth = GuessProxyAuth(None,None,"user","passwd")
```
Comment (sigmavirus24) [inline tests/test_auth.py]: Please add spaces after commas, e.g., 

```py
GuessProxyAuth(None, None, "user", "passwd")
```

File: tests/test_auth.py [hunk 11]
```diff
@@ -44,3 +45,36 @@ def test_no_auth(self):
...
+        guess_auth.handle_407(r)
+
+        mock_handle_407.assert_called_with(r)
+
+
+    @mock.patch('requests.auth.HTTPProxyAuth.__call__')
+    @mock.patch('requests.cookies.extract_cookies_to_jar')
+    def test_handle_407_header_basic(self, mock_cookiejar, mock_proxyAuth_call):
```
Comment (sigmavirus24) [inline tests/test_auth.py]: Can we just rename these to `extract_cookies_to_jar` and `proxy_auth_call` instead of `mock_*`?

File: tests/test_auth.py [hunk 12]
```diff
@@ -44,3 +45,36 @@ def test_no_auth(self):
...
+        mock_proxyAuth_call.return_value = requests.Response()
+
+        kwargs = {}
+        r.headers['Proxy-Authenticate'] = 'Basic realm="Fake Realm"'
+        guess_auth = GuessProxyAuth(None,None,"user","passwd")
+        guess_auth.handle_407(r, *kwargs)
+
+        mock_proxyAuth_call.assert_called_with(req)
+
+
```
Comment (sigmavirus24) [inline tests/test_auth.py]: Could you trim these trailing new lines too?

**Comment by sigmavirus24** [comment]
A couple final nits, otherwise :+1:

**Comment by pcreech** [comment]
@sigmavirus24 and those should now be taken care of!

**Comment by sigmavirus24** [comment]
Thanks @pcreech !
